JVM IR: Fix handling of anonymous initializers in inline classes (KT-46554)

sfs opened the pull request.

udalov commented:
Is a more localized fix possible maybe? This one broke the compiler bootstrap in the following case, which makes me a bit wary:
```
class A(c: C, xs: List<Long>) {
    init {
        xs.forEach { c.f(it) }
    }
}

@JvmInline
value class Z(val x: Long)

class C {
    fun f(x: Long): Z = Z(x)
}

fun main() {
    A(C(), listOf(1L))
}
```
->
```
Exception in thread "main" java.lang.NoSuchMethodError: C.f(J)J
	at A.<init>(1.kt:3)
	at _1Kt.main(1.kt:15)
	at _1Kt.main(1.kt)
```

sfs commented:
That's what I get for making an "obvious refactoring" right before opening a PR. It's fixed now.

As for a more localized fix, let me explain what's going on with the code. It turns out that there were two issues here.

Inside of the primary constructor of an inline class we don't have `<this>` in scope - rather we have an unboxed version of the class in scope as the first `argument` to the primary constructor. What the previous code then tried to do was replace accesses to `<this>` with `<unsafe-coerce>(argument)`. This is necessary in order for the types to match and the rest of the transformation to work correctly. One problem was that the code detected this situation by looking at the surrounding scopes, which breaks as soon as you start nesting functions or classes inside of init blocks, e.g., code like this:
```kotlin
var result = "Fail"

inline class A(val value: String) {
    init {
        class B {
            init {
                result = value
            }
        }
        B()
    }
}
```
also failed with an exception. The new code in this PR is simpler: it introduces a temporary variable to hold the value of `<unsafe-coerce>(argument)` and uses the existing code (=> `valueMap`) in the lowering to replace variable accesses.

Now at least the reduced test case for KT-46554 actually fails because of a different issue, which is that anonymous initializers (and default arguments in the primary constructor) were visited twice. Once when the lowering processed the body of the class and then again in `buildPrimaryInlineClassConstructor`. The lowering is really set up to only process things once. In this example the issue was that the `valueMap` is not scoped, so if you process a secondary constructor before an init block you end up replacing `<this>` with the wrong variable and that caused a problem during codegen. In fact, there are more problems when processing things twice, such as function calls being mangled twice, which is what the additional tests were for.

Incidentally, this is also what I refactored before submitting the PR. Previously my code processed anonymous initializers as part of the body of the class and just replaced the variable in `buildPrimaryInlineClassConstructor`. On the other hand, since we can do the replacement at the same time as the traversal it makes more sense to only process initializers in `buildPrimaryInlineClassConstructor`. That's when I stubbed out `visitAnonymousInitializer` ... for everything, not just inline classes. And that's why the code was broken. It's not that this is a subtle change, it was just me being stupid.

Anyway, the code is fixed now and I've added a new test for this problem (`testInlineClassInInitBlock`). Would be great if you could take another look, since I don't see how to fix these problems in a more localized fashion.

---

And yes, before you ask, this lowering could do with some more comments. It's on my todo list now.

udalov commented:
Thanks for the elaborate explanation! Yeah, it makes total sense now and I'm pretty sure that it was never intended to visit the same IR twice :). Merged manually.
